### ISSUE TITLE
Support for MSVC++

### ISSUE THREAD

**Issue by plicease** [plicease/FFI-Platypus#16]
This requires adding support for MSVC++ to `Alien::FFI`.

This should be POSSIBLE since I managed to add MSVC++ support to `FFI::Raw`, though it won't be pretty.

**Comment by plicease** [comment]
Another aspect of this is going to be writing typedefs for uint*_t types.  That shouldn't be too hard, Math::Int64 has some examples for int64_t, ther other sizes should follow pretty easy.

**Comment by plicease** [comment]
Here is the relevant PR to accomplish this for `FFI::Raw`
https://github.com/ghedo/p5-FFI-Raw/pull/19

**Comment by plicease** [comment]
This is not working, but was able to compile (not link) with this diff:

```
xian-x86_64% git diff
diff --git a/inc/My/ModuleBuild.pm b/inc/My/ModuleBuild.pm
index d58d297..c05c13c 100644
--- a/inc/My/ModuleBuild.pm
+++ b/inc/My/ModuleBuild.pm
@@ -32,6 +32,7 @@ sub new
   if($^O eq 'MSWin32' && $Config{cc} =~ /cl(\.exe)?$/i)
   {
     $args{extra_linker_flags} .= ' psapi.lib';
+    $args{extra_compiler_flags} = '-DFFI_BUILDING ' . $args{extra_compiler_flags};
   }
   elsif($^O =~ /^(MSWin32|cygwin)$/)
   {
diff --git a/include/ffi_platypus.h b/include/ffi_platypus.h
index 903a546..f73ad3c 100644
--- a/include/ffi_platypus.h
+++ b/include/ffi_platypus.h
@@ -79,7 +79,7 @@ typedef struct _ffi_pl_type_extra_closure {
   ffi_cif ffi_cif;
   int flags;
   struct _ffi_pl_type *return_type;
-  struct _ffi_pl_type *argument_types[0];
+  struct _ffi_pl_type *argument_types[1];
 } ffi_pl_type_extra_closure;

 typedef union _ffi_pl_type_extra {
@@ -91,7 +91,7 @@ typedef union _ffi_pl_type_extra {
 typedef struct _ffi_pl_type {
   ffi_type *ffi_type;
   platypus_type platypus_type;
-  ffi_pl_type_extra extra[0];
+  ffi_pl_type_extra extra[1];
 } ffi_pl_type;

 typedef struct _ffi_pl_function {
@@ -99,7 +99,7 @@ typedef struct _ffi_pl_function {
   void *platypus_sv;  /* really a Perl SV* */
   ffi_cif ffi_cif;
   ffi_pl_type *return_type;
-  ffi_pl_type *argument_types[0];
+  ffi_pl_type *argument_types[1];
 } ffi_pl_function;

 typedef struct _ffi_pl_closure {
diff --git a/xs/windl.c b/xs/windl.c
index 994f897..316bd4b 100644
--- a/xs/windl.c
+++ b/xs/windl.c
@@ -33,7 +33,7 @@ static const char *error = NULL;
 void *
 windlopen(const char *filename, int flags)
 {
-  ssize_t size;
+  size_t size;
   char *win_path_filename;
   library_handle *handle;

```

**Comment by plicease** [comment]
Interesting, with Visual C++, this won't compile:

```
typedef struct _bar { int z; int xx[0]; } bar;
typedef struct _foo { int x; bar y[0]; } foo;
```

but this will:

```
typedef struct _bar { int z; int xx[1]; } bar;
typedef struct _foo { int x; bar y[0]; } foo;
```

*sigh*

**Comment by plicease** [comment]
Here is the linker error I am (currently) seeing:

```
ffi.LIB(ffi.o) : error LNK2019: unresolved external symbol __alloca referenced in function _ffi_closure_WIN32_inner
C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\sysv.exe : fatal error LNK1120: 1 unresolved externals
Generating script 'C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.ccs'
cl -nologo -c @"C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.ccs" -Fo"C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.obj" "C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.c"
thiscall.c
link /OUT:C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.exe C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.obj /LIBPATH:C:/PROGRA~2/PERLAL~2/libffi/lib/PKGCON~1/../../lib ffi.LIB psapi.lib
Microsoft (R) Incremental Linker Version 12.00.21005.1
Copyright (C) Microsoft Corporation.  All rights reserved.

ffi.LIB(ffi.o) : error LNK2019: unresolved external symbol __alloca referenced in function _ffi_closure_WIN32_inner
C:\Users\ollisg\AppData\Local\Temp\6Gw_5HDvzS\thiscall.exe : fatal error LNK1120: 1 unresolved externals
```

This is with a libffi cross compiled from my linux box.  The same libffi binary works with strawberry.

**Comment by plicease** [comment]
I've not been able to get a clean build of `libffi` from the current version of `Alien::FFI` with MSVC++ Perl.  At one point this was working with an older version of `libffi` in `FFI::Raw`.

**Comment by plicease** [comment]
Ah!  Built a DLL for `libffi` and this seems to work better, although now I will need to make sure that the DLL is in the path.   With the above commit, and the dll in the path it builds and passes all of the tests!

**Comment by plicease** [comment]
Here are some details on alloca on MSVC++:

https://msdn.microsoft.com/en-us/library/5471dc8s.aspx

Note the page seems to lock the browser due to some badly behaved javascript.

**Comment by plicease** [comment]
_malloca

https://github.com/plicease/FFI-Platypus/commit/2b077f67a383dd353e930632dfaaf8ed376a3a0e

**Comment by plicease** [comment]
This seems to fix source build for `Alien::FFI`:

https://github.com/plicease/Alien-FFI/commit/b4a8ae3bdef007973ddfd46492ee075c62317b4c

**Comment by plicease** [comment]
This is fixed in 0.32
